### PR TITLE
Add seastar signal api

### PR DESCRIPTION
**Pull request by tomershafir**
fixes: https://github.com/scylladb/seastar/issues/261.

The PR adds an object based interface for signals on seastar namespace.

- Adds `signal.hh` header implemented by `reactor.cc`.
- Adds `doc/signal.md`.
- Deprecates `reactor::handle_signal`.
- Supports C++20 modules.

Future possibilities:
- Add signal handler unset logic, and unset signal handler on destructor.
- Consider to add a reliable test case for `once` (should require reactor changes to expose an indicator).

### CODE REVIEW

**Comment by xemul** [comment]
@tomershafir , please fix the commits sequence, it contains some "Merge branch 'master'" and a post-fixlet, both are not welcome

File: doc/signal.md [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+Seastar provides an interface to handle signals natively and safely as asynchronous tasks.
+
+It provides a dedicated header called [seastar/core/signal.hh](../include/seastar/core/signal.hh).
+
+## Usage
+
+You need to construct a `signal_handler` instance that will register the provided signal handler for the specified signal based on the configuration params.
+
+At the moment, the signal handler will remain past object destruction.
```
Comment (xemul) [inline doc/signal.md:11]: That's extremely confusing. I would expect that desctruction of an object unregisters the handler

File: include/seastar/core/reactor.hh [hunk 2]
```diff
@@ -354,13 +354,23 @@ private:
...
     /// Register a user-defined signal handler
+    [[deprecated("Use seastar::signal_handler sh(signo, handler); instead")]]
     void handle_signal(int signo, noncopyable_function<void ()>&& handler);
     void wakeup();
     /// @private
     bool stopped() const noexcept { return _stopped; }
 
 private:
     class signals {
+        struct _signal_handler {
```
Comment (xemul) [inline include/seastar/core/reactor.hh:365]: Leading underscore is for private members, class names shouldn't have it

File: include/seastar/core/reactor.hh [hunk 3]
```diff
@@ -371,20 +381,15 @@ private:
...
     };
-
+    
     signals _signals;
     std::unique_ptr<thread_pool> _thread_pool;
+    
     friend class thread_pool;
     friend class thread_context;
     friend class internal::cpu_stall_detector;
+    friend struct signal_handler;
```
Comment (xemul) [inline include/seastar/core/reactor.hh:392]: New friends for reactor class are not welcome. It's large enough already

Comment (tomershafir) [reply include/seastar/core/reactor.hh:392]: The other ways would be:

1. Add public `reactor::handle_signal_once` and undo deprecation of `reactor::handle_signal`. Users will be able to call them and I think it inverts the purpose of changing the api. Also, it adds more code to the reactor.
2. Refactor signal logic out of the reactor. It seems like the blast radius is big and it's not an easy task.

Thus I would prefer the `friend struct` solution at the moment. Let me know wdyt

Comment (xemul) [reply include/seastar/core/reactor.hh:392]: We already have nested class signal, public nested class signal_handler would be nice replacement for explicit friend. But as I wrote above, installing handler in constructor smells like RAII and implies unsetting it in destructor

Comment (tomershafir) [reply include/seastar/core/reactor.hh:392]: My goal has been to add a clean seastar namespace level api, not through the reactor.

File: doc/signal.md (hunk 1 above)
Comment (tomershafir) [reply doc/signal.md:11]: Ill add unregister logic to this PR

**Comment by tomershafir** [comment]
@xemul Regarding destructor logic, it seems to me like the RAII approach here may be less suitable. As the POSIX api doesn't fit RAII naturally, adapting it can introduce complexity and confusion. I assume that multiple objects can be constructed for a single `signo`, like POSIX.

For example:
```C++
seastar::signal_handler signal_handler1(SIGINT, [this] { signaled(); });
...
seastar::signal_handler signal_handler2(SIGINT, [this] { signaled(); });
...
```
What should happen if `signal_handler1` is destroyed after `signal_handler2` is constructed and before it's destroyed?
What should happen if `signal_handler1` is destroyed after `signal_handler2` is destroyed?
The possible solutions to these problems don't seem natural for RAII to me.
 
I think I should re-target this PR to introduce a clean seastar namespace level api, and do it via procedures rather than objects. Wdyt?

**Comment by xemul** [comment]
> I think I should re-target this PR to introduce a clean seastar namespace level api, and do it via procedures rather than objects. Wdyt?

Yes, this sounds good

**Comment by tomershafir** [comment]
@xemul I force pushed the changes

File: tests/unit/signal_test.cc [hunk 4]
```diff
@@ -20,6 +20,7 @@
  */
 
 #include <seastar/core/reactor.hh>
```
Comment (xemul) [inline tests/unit/signal_test.cc:22]: Is this header still needed?

File: include/seastar/core/reactor.hh [hunk 5]
```diff
@@ -371,21 +380,17 @@ private:
...
-            noncopyable_function<void ()> _handler;
-        };
-        std::atomic<uint64_t> _pending_signals;
-        std::unordered_map<int, signal_handler> _signal_handlers;
     };
-
+    
     signals _signals;
     std::unique_ptr<thread_pool> _thread_pool;
+
```
Comment (xemul) [inline include/seastar/core/reactor.hh:387]: Unrelated change

File: include/seastar/core/reactor.hh [hunk 6]
```diff
@@ -371,21 +380,17 @@ private:
...
-    private:
-        struct signal_handler {
-            signal_handler(int signo, noncopyable_function<void ()>&& handler);
-            noncopyable_function<void ()> _handler;
-        };
-        std::atomic<uint64_t> _pending_signals;
-        std::unordered_map<int, signal_handler> _signal_handlers;
     };
-
+
```
Comment (xemul) [inline include/seastar/core/reactor.hh:384]: Unwanted indentation

File: include/seastar/core/reactor.hh [hunk 7]
```diff
@@ -371,21 +380,17 @@ private:
         void handle_signal_once(int signo, noncopyable_function<void ()>&& handler);
         static void action(int signo, siginfo_t* siginfo, void* ignore);
         static void failed_to_handle(int signo);
-    private:
-        struct signal_handler {
-            signal_handler(int signo, noncopyable_function<void ()>&& handler);
-            noncopyable_function<void ()> _handler;
-        };
-        std::atomic<uint64_t> _pending_signals;
-        std::unordered_map<int, signal_handler> _signal_handlers;
```
Comment (xemul) [inline include/seastar/core/reactor.hh:380]: Why is this block moved upwards?

File: src/core/signal.cc [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+module seastar;
+#else
+#include <seastar/core/signal.hh>
+#include <seastar/core/reactor.hh>
+#endif
+
+namespace seastar {
+
+static reactor& engine_or_throw() {
+    if (!engine_is_ready()) {
```
Comment (xemul) [inline src/core/signal.cc:35]: I think that's impossible. No tasks can be run until engine is ready

File: include/seastar/core/reactor.hh (hunk 7 above)
Comment (tomershafir) [reply include/seastar/core/reactor.hh:380]: To remove redundant `private` in a class

Comment (tomershafir) [reply include/seastar/core/reactor.hh:380]: Ill extract it into a different PR

File: tests/unit/signal_test.cc (hunk 4 above)
Comment (tomershafir) [reply tests/unit/signal_test.cc:22]: No, will remove

File: src/core/signal.cc (hunk 8 above)
Comment (tomershafir) [reply src/core/signal.cc:35]: Its possible, if a user calls the procedure before `app.run()`, like:

```C++
seastar::handle_signal(SIGINT, [] {});
return app.run(argc, argv, [&app] -> seastar::future<> {
...
```

I dont think its obvious that it must be called inside `run()`, so I think the logic provides a more controlled and understandable error handling.

Comment (xemul) [reply src/core/signal.cc:35]: In this case the "reactor must be initialized" can be confusing message, because it's not obvious that "app.run()" initializes the reactor. I would just mention in the method doc, that it's only valid to use one from inside app.run() lambda, otherwise it's UB

File: src/core/signal.cc [hunk 9]
```diff
@@ -31,15 +31,8 @@ module seastar;
 
 namespace seastar {
 
-static reactor& engine_or_throw() {
```
Comment (xemul) [inline src/core/signal.cc]: Please, squash it with the previous commit

Comment (tomershafir) [reply src/core/signal.cc]: sorry, done

**Comment by tomershafir** [comment]
@xemul close https://github.com/scylladb/seastar/issues/261?
